### ISSUE TITLE
Support attachment of folders

### ISSUE THREAD

**Issue by adubovik** [epam/ai-dial-chat#967]
> [!WARNING]
> The design in this comment was rejected in favour of the one described in the follow-up reply.

### Description

Currently only individual files could be attached to a user message. 

However, it's useful for RAG-like application to accept the whole folders from the user workspace.

Attaching dozens of individual files could be very inconvenient.

Moreover, we want an easy way to identify collections of files for which RAG applications build indices. The specific folder in a user workspace could be such an identifier.

Whether application supports folder attachments or not, should be defined by a feature flag from the listing `features.folder_attachments`.

There is no MIME type for folders, but we have to provide one in `attachments[*].type` field. 
The most appropriate type would be the catch-all type `*/*`, since a folder may contain all types of files, or `null`.

### Related issues

Backend:
* Support auto-sharing of folders: https://github.com/epam/ai-dial-core/issues/265
* Feature flag in the core listing: https://github.com/epam/ai-dial-core/issues/266

**Comment by sdryapko** [comment]
We are going to introduce new MIME type "application/vnd.dial.folder-metadata+json" and use it to attach folders (via PUT query) with url equals to path to the folder. 

Same MIME type will be returned as a response when you are trying to do GET for same url for same folder and it will return to you folder metadata with all folder structure in it including all files (for now without pagination but pagination can be implemented in the future because we already have it designed for this kind of API).

**Comment by adubovik** [comment]
# Requirements

1. It should be possible to pass to a DIAL application the whole folder from a user workspace instead of individual files.
2. Whatever files the given folder contains should be accessible to the application in the same way as if they were passed individually.

## Metadata MIME type proposal

The current proposal suggests to introduce a new MIME type for DIAL metadata calls: `application/vnd.dial.metadata+json`

## Implications for chat completion attachment API

1. Attachment may point to a DIAL **folder** using the new MIME type.
2. Attachment may point to a DIAL **files** using the type too. Which gives us two ways to address a file:
    1. by value (`files/BUCKET/PATH` + `Document MIME type`)
    2. by reference (`metadata/files/BUCKET/PATH` + `Metadata MIME type`)
    The first way is preferable, because it doesn't require addition call to get the file content type.

## Implications for DIAL Chat

Chat should support **folder** attachments following this format:

```json
{
  "type": "application/vnd.dial.metadata+json",
  "url": "metadata/files/BUCKET/foo/baz/folder/"
}
```

Whether an application supports folder attachments or not must be determined by `features.folderAttachmentsSupported` field from the Core listing. By default the field is `false`:

```json
"features": {
  "folderAttachmentsSupported": true
}
```

## Implications for DIAL application developers

1. The DIAL application supporting **folder** attachments should be able to handle the new MIME type.
2. The DIAL application supporting **file** attachment should be able to support **both** ways of addressing a file.

In general, applications must perform input data validation and only accept attachment type, which they are able to handle and reject the rest.

## Implications for DIAL Core

1. Metadata call should return `Content-Type:application/vnd.dial.metadata+json` for `GET metadata/RESOURCE_TYPE/RESOURCE_ID` calls given that `Accept` requested this type.

    In order to preserve backward compatibility with existing metadata client, the following content-type mapping should be used:

    |Accept          |Content-Type    |
    |----------------|----------------|
    |`*/*`             |`application/json`|
    |`application/json`|`application/json`|
    |`application/vnd.dial.metadata+json`|`application/vnd.dial.metadata+json`|

    This arguably defeats the purpose of introducing a new MIME type, but it's a necessary step to ensure backward compatibility.

2. Access control (auto-sharing) for **files** and **folders** must be adjusted to support DIAL metadata references.

See the issue https://github.com/epam/ai-dial-core/issues/313

## Implications for DIAL Core config

`folderAttachmentsSupported` field in the Core config declares if an application supports folder attachments:

```json
"features": {
  "folderAttachmentsSupported": true
}
```

Note that we don't put any constraints on the content of the folder. It's application's responsibility to handle the folder content: select supported documents and complain or ignore those it doesn't support.

What about `inputAttachmentTypes` config field, which declares MIME types of supported attachment types?

1. `inputAttachmentTypes` field only concerns **file** attachments. It tells that an application supports file attachments with the given MIME types and URLs pointing directly to these documents.
2. `features.folderAttachmentsSupported` field only concerns DIAL **folder** attachments. It means that an application supports attachments with the DIAL metadata MIME type and URL pointing to a DIAL folder.

So these fields are independent and should be handled separately. For example:

1. Config of an application supporting only **folder** attachments and no **file** attachments:

    ```json
    {
      "inputAttachmentTypes": [],
      "features": {
        "folderAttachmentsSupported": true
      }
    }
    ```

2. Config of an application supporting only **file** PDF attachments and no **folder** attachments:

    ```json
    {
      "inputAttachmentTypes": ["application/pdf"],
      "features": {}
    }
    ```

3. Config of an application supporting arbitrary **file** attachments and **folder** attachments:

    ```json
    {
      "inputAttachmentTypes": ["*/*"],
      "features": {}
    }
    ```

4. Config of an application supporting neither **file** attachments not **folder** attachments. It's an empty config, because by default these features are disabled:

    ```json
    {}

## Attachment examples

File attachment via direct reference:

```json
{
  "type": "text/plain",
  "url": "files/BUCKET/foo/baz/folder/doc.txt"
}
```

File attachment via metadata:

```json
{
  "type": "application/vnd.dial.metadata+json",
  "url": "metadata/files/BUCKET/foo/baz/folder/doc.txt"
}
```

Folder attachment via metadata:

```json
{
  "type": "application/vnd.dial.metadata+json",
  "url": "metadata/files/BUCKET/foo/baz/folder/"
}
```

## Misc

### Trailing slash matters for metadata calls

```json
GET metadata/files/BUCKET/a/b/c
===
{
  "name": "c",
  "parentPath": "a/b",
  "bucket": "BUCKET",
  "url": "files/BUCKET/a/b/c/",
  "nodeType": "FOLDER",
  "resourceType": "FILE",
  "items": null
}
```

```json
GET metadata/files/BUCKET/a/b/c/
===
{
  "name": "c",
  "parentPath": "a/b",
  "bucket": "BUCKET",
  "url": "files/BUCKET/a/b/c/",
  "nodeType": "FOLDER",
  "resourceType": "FILE",
  "items": [
    {
      "name": "d.txt",
      "parentPath": "a/b/c",
      "bucket": "BUCKET",
      "url": "files/BUCKET/a/b/c/d.txt",
      "nodeType": "ITEM",
      "resourceType": "FILE",
      "contentLength": 60,
      "contentType": "text/plain"
    }
  ]
}
```

### Paging is not supported for files/folders

It's currently supported only for **prompts** and **conversations** resources.

### There is no PUT request for folder upload

We upload files one at a time.

## FAQ

Q: Why not use std MIME type `application/json` for metadata calls? After all `GET metadata/files/...` already returns `Content-Type:application/json`.
A: Wherever a diverse set of documents is expected to be handled in a uniform way, it is better to use a custom media type, so that the client can distinguish between different types of attachments and handle them accordingly. `application/json` is too generic, whereas, the custom type tells that the attachment represents DIAL folder and, thus, client may decide to handle/render it in a specific way.

**Comment by sdryapko** [comment]
@IlyaBondar  please look into it

**Comment by VolhaBazhkova** [comment]
It was verified and released in 0.11 version
